packagedb/find_source_repo.py

```diff
@@ -0,0 +1,454 @@
+    """
+    Return type for get_urls_from_text
+    """
+
+    url = "url"
+    text = "text"
+
+
+non_reachable_urls = []
+CACHE = {
+    #    url: data
+}
+
+
+def get_urls_from_text(text):
+    """
+    Return the URLs found in a text
+    """
+    lines = text.splitlines()
+    # location can be a list of lines
+    for url in get_urls_from_location(location=lines)["urls"]:
+        yield url["url"]
+
+
+def get_data_from_response(response, data_type=URLDataReturnType.text):
+    """
+    Return the data from a response
+    """
+    if not response:
+        return
+    data_by_type = {
+        URLDataReturnType.url: response.url,
+        URLDataReturnType.text: response.text,
+    }
+    if data_type in data_by_type:
+        return data_by_type[data_type]
+    else:
+        raise ValueError(f"Invalid data_type: {data_type}")
+
+
+def get_data_from_url(
+    url,
+    data_type=URLDataReturnType.text,
+    timeout=None,
+):
+    """
+    Return the text from URL
+    """
+    try:
+        if not url:
+            return
+        if url in non_reachable_urls:
+            return
+        if url in CACHE:
+            response = CACHE[url]
+            return get_data_from_response(response=response, data_type=data_type)
+        response = requests.get(url=url, timeout=timeout)
+        if response.status_code != 200:
+            non_reachable_urls.append(url)
+            return
+        CACHE[url] = response
+        return get_data_from_response(response=response, data_type=data_type)
+    except Exception as e:
+        logger.error(f"Error getting {url}: {e}")
+        non_reachable_urls.append(url)
+        return
+
+
+def convert_apache_svn_to_github_url(url):
+    """
+    Convert an SVN URL to a GitHub URL
+    >>> convert_apache_svn_to_github_url("svn+https://svn.apache.org/repos/asf/cdi/tags/1.0/cdi-extension-el-jsp")
+    'https://github.com/apache/cdi/tree/1.0'
+    """
+    # svn+https://svn.apache.org/repos/asf/{name}/tags/{tag}/cdi-extension-el-jsp
+    if "svn+" in url:
+        _, _, url = url.partition("svn+")
+    urlparsed_result = urlparse(url=url)
+    # https://github.com/apache/{name}/tree/{tag}
+    if urlparsed_result.netloc != "svn.apache.org":
+        return
+    path = urlparsed_result.path
+    path_segs = path.split("/")
+    if len(path_segs) < 6:
+        return
+    if path_segs[1] != "repos":
+        return
+    if path_segs[2] != "asf":
+        return
+    if path_segs[4] != "tags":
+        return
+    name = path_segs[3]
+    tag = path_segs[5]
+    return f"https://github.com/apache/{name}/tree/{tag}"
+
+
+def add_source_repo_to_package_set(
+    source_repo_type,
+    source_repo_name,
+    source_repo_namespace,
+    source_repo_version,
+    download_url,
+    purl,
+    source_purl,
+    package,
+):
+    """
+    Take source repo package information, create source package
+    and add it to a Package set
+    """
+    package_set = package.package_sets.first()
```

Comment:
I think we will want to add the source_repo_package to every PackageSet this Package is part of.

@pombredanne Do we want to add the source repo package to all package sets of a package for all package_content types?

Reply:
Changed the code to add to all package sets.

Reply:
> Do we want to add the source repo package to all package sets of a package for all package_content types?

I think for now yes.

We should discuss further and refactor later as we experience how this works in practice.
Right now package set are based on one base binary. For instance if I have 50 wheels and one sdist for a pypi package we will have 50 sets each with one wheel and the sdist. I think this is the more flexible approach. We may in some cases have only one set for all the binaries and the one source. That is also possible with the current model AFAICS, and we may elect to use this for some packages in the future.